drivers/console/aarch64/skeleton_console.S

```diff
@@ -6,97 +6,179 @@
-	mov	w0, #1
-	ret
-core_init_fail:
-	mov	w0, wzr
+func console_xxx_register
+	/*
+	 * Store parameters (e.g. hardware base address) in driver-specific
+	 * console_xxx_t structure field if they will need to be retrieved
+	 * by later console callback (e.g. putc).
+	 * Example:
+	 */
+	str	x1, [x0, #CONSOLE_T_XXX_BASE]
+	str	x2, [x0, #CONSOLE_T_XXX_SOME_OTHER_VALUE]
+
+	/*
+	 * Initialize console hardware, using x1 - x7 parameters as needed.
+	 * Keep console_t pointer in x0 for later.
+	 */
+
+	/*
+	 * Fill out remaining console_t fields, including default flags. Set
+	 * function callbacks to NULL if they don't apply to this driver.
+	 * Example:
+	 */
+	adrp	x1, console_xxx_putc
+	add	x1, x1, :lo12:console_xxx_putc
+	str	x1, [x0, #CONSOLE_T_PUTC]
+	adrp	x1, console_xxx_getc
+	add	x1, x1, :lo12:console_xxx_getc
+	str	x1, [x0, #CONSOLE_T_GETC]
+	mov	x1, #0
+	str	x1, [x0, #CONSOLE_T_FLUSH]
+	mov	x1, #(1 << CONSOLE_FLAG_BOOT_SHIFT | \
+		      1 << CONSOLE_FLAG_CRASH_SHIFT)
+	str	x1, [x0, #CONSOLE_T_FLAGS]
+
+	/* Tail call console_register, which has the same return signature. */
+	b	console_register
+	
+	/* Jump here if hardware init fails or parameters are invalid. */
+register_fail:
+	mov	w0, #0
 	ret
-endfunc console_core_init
+endfunc console_xxx_register
 
 	/* --------------------------------------------------------
-	 * int console_core_putc(int c, uintptr_t base_addr)
+	 * int console_xxx_putc(int c, console_xxx_t *console)
 	 * Function to output a character over the console. It
 	 * returns the character printed on success or -1 on error.
 	 * In : w0 - character to be printed
-	 *      x1 - console base address
-	 * Out : return -1 on error else return character.
-	 * Clobber list : x2
+	 *      x1 - pointer to console_t struct
+	 * Out: w0 - printed character on success, < 0 on error.
+	 * Clobber list : x0, x1, x2
 	 * --------------------------------------------------------
 	 */
-func console_core_putc
-	/* Check the input parameter */
-	cbz	x1, putc_error
-	/* Insert implementation here */
+func console_xxx_putc
+	/*
+	 * Retrieve values we need (e.g. hardware base address) from
+	 * console_xxx_t structure pointed to by x1.
+	 * Example:
+	 */
+	ldr	x1, [x1, #CONSOLE_T_XXX_BASE]
+
+	/*
+	 * Write w0 to hardware.
+	 */
+
 	ret
+
+	/* Jump here if output fails for any reason. */
 putc_error:
 	mov	w0, #-1
 	ret
-endfunc console_core_putc
+endfunc console_xxx_putc
 
 	/* ---------------------------------------------
-	 * int console_core_getc(uintptr_t base_addr)
+	 * int console_xxx_getc(console_xxx_t *console)
 	 * Function to get a character from the console.
-	 * It returns the character grabbed on success
-	 * or -1 on error.
-	 * In : x0 - console base address
+	 * It returns the character grabbed on success,
+	 * -1 if no character available, or any value
+	 * between -2 and -127 otherwise
+	 * In : x0 - pointer to console_t struct
+	 * Out: w0 - character on success, -1 if no char
+	             available, < -1 on error
```

Comment:
> So you're introducing a policy that drivers should now not block but return success if they have a char ready in their FIFO, so that the char to return is from the first "ready" driver. I don't see any real need for this change. Drivers could continue blocking until they have a char (or a real error).

But... I don't see how this could work when you have more than one console. Let's say you're receiving characters on the console that comes second in the list... console_getc() would call getc() from the console that's first in the list, and then that one would just block forever (waiting for characters that will never come) while your input on the second console never gets handled. I think the only way you can really make a blocking getc() work across multiple consoles is to make the drivers non-blocking and keep looping over them like I have it here.

> Also this relies on all drivers being updated to the new policy for it to work properly (which isn't yet done).

My patches to the individual console drivers make the getc() implementations non-blocking so they work with the new API.

Reply:
> console_getc() would call getc() from the console that's first in the list, and then that one would just block forever (waiting for characters that will never come) while your input on the second console never gets handled.

Yes, that would have been acceptable for now. However, given that you've already made this change, and it looks like this is backwards compatible for existing drivers (i.e. they will still work if they are not changed to be non-blocking), you can leave it as it is now.

> My patches to the individual console drivers make the getc() implementations non-blocking so they work with the new API.

Sorry, I missed that.

Reply:
> it looks like this is backwards compatible for existing drivers

Existing drivers (e.g. in some non-upstream platform directory) won't be compatible with MULTI_CONSOLE_API at all anyway -- neither for getc, nor for putc. They're supposed to keep using the legacy API until the platform as a whole is updated to use the new API... and when that happens, the driver should also be updated to do a non-blocking getc. (I mean, yeah, it will still kinda work without that but the idea is that non-blocking is the correct way to implement the new API. Should be a simple change for most drivers anyway.)

Reply:
Yes, agree.